dataflow/gemma-flex-template/README.md

```diff
@@ -0,0 +1,179 @@
+
+To customize the behavior of the handler, implement the following methods: `load_model`, `validate_inference_args`, and `share_model_across_processes`.
+
+The PyTorch implementation of the Gemma models has a `generate` method
+that generates text based on a prompt. To route the prompts correctly, use this function in the `run_inference` function.
+
+```py
+class GemmaPytorchModelHandler(ModelHandler[str, PredictionResult,
+                                            GemmaForCausalLM]):
+    def __init__(self,
+                 model_variant: str,
+                 checkpoint_path: str,
+                 tokenizer_path: str,
+                 device: Optional[str] = 'cpu'):
+        """ Implementation of the ModelHandler interface for Gemma-on-Pytorch
+        using text as input.
+
+        Example Usage::
+
+          pcoll | RunInference(GemmaPytorchHandler())
+
+        Args:
+          model_variant: The Gemma model name.
+          checkpoint_path: the path to a local copy of gemma model weights.
+          tokenizer_path: the path to a local copy of the gemma tokenizer
+          device: optional. the device to run inference on. can be either
+            'cpu' or 'gpu', defaults to cpu. 
+        """
+        model_config = get_config_for_2b(
+        ) if "2b" in model_variant else get_config_for_7b()
+        model_config.tokenizer = tokenizer_path
+        model_config.quant = 'quant' in model_variant
+        model_config.tokenizer = tokenizer_path
+
+        self._model_config = model_config
+        self._checkpoint_path = checkpoint_path
+        if device == 'GPU':
+            logging.info("Device is set to CUDA")
+            self._device = torch.device('cuda')
+        else:
+            logging.info("Device is set to CPU")
+            self._device = torch.device('cpu')
+        self._env_vars = {}
+
+    def load_model(self) -> GemmaForCausalLM:
+        """Loads and initializes a model for processing."""
+        torch.set_default_dtype(self._model_config.get_dtype())
+        model = GemmaForCausalLM(self._model_config)
+        model.load_weights(self._checkpoint_path)
+        model = model.to(self._device).eval()
+        return model
+```
+
+### Formatting DoFn
+
+The output from a keyed model handler is a tuple of the form `(key, PredictionResult)`. To format that output into a string before sending it to the answer Pub/Sub topic, use an extra `DoFn`.
+
+```py
+class FormatOutput(beam.DoFn):
+  def process(self, element, *args, **kwargs):
+    yield "Key : {key}, Input: {input}, Output: {output}".format(key=element[0], input=element[1].example, output=element[1].inference)
+```
+
+## Build the Flex Template
+Run the following code from the directory to build the Dataflow flex template.
+
+- Replace `$TEMPLATE_FILE` with a Google Cloud Storage path written as a JSON file.
```

Comment:
This was unclear to me when going through these instructuons. 

The $TEMPLATE_FILE needs to be a GCS storage path, that will be used as the output for this command. The "written as a JSON file" denotes it should be a `.json` file, but given this description and the section description, I wasn't sure if this was a file I had to create with contents relating to the GCS bucket. 

A way to make this more clear might be to specify "Create a bucket" then use that value as part of the variable: `gs://$CONFIG_BUCKET/config.json` (for example).

Reply:
You also never explicitly say to create any buckets. This is possibly one of many you need (one for the config, possibly a separate one for staging later)

Reply:
added instructions for creating a bucket, altered instructions for the template file path